File: Java-Core/src/main/java/presto/type/IntegerType.java
package presto.type;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.grammar.Identifier;
import presto.runtime.Context;
import presto.value.ICollection;
import presto.value.IValue;
import presto.value.Integer;
import presto.value.IntegerRange;
import presto.value.Range;
import presto.value.ListValue;

public class IntegerType extends NativeType {

	static IntegerType instance = new IntegerType();
	
	public static IntegerType instance() {
		return instance;
	}
	
	private IntegerType() {
		super("Integer");
	}
	
	@Override
	public Class<?> toJavaClass() {
		return Long.class;
	}
	
	@Override
	public boolean isAssignableTo(Context context, IType other) {
		return (other instanceof IntegerType) || (other instanceof DecimalType) || (other instanceof AnyType);
	}
	
	@Override
	public IType checkAdd(Context context, IType other, boolean tryReverse) throws SyntaxError {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return other;
		return super.checkAdd(context, other, tryReverse);
	}
	
	@Override
	public IType checkSubstract(Context context, IType other) throws SyntaxError {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return other;
		return super.checkSubstract(context, other);
	}
	
	@Override
	public IType checkMultiply(Context context, IType other, boolean tryReverse) throws SyntaxError {
		if(other instanceof IntegerType)
			return this;
		if(other instanceof DecimalType)
			return other;
		if(other instanceof CharacterType)
			return TextType.instance();
		if(other instanceof TextType)
			return other;
		if(other instanceof PeriodType)
			return other;
		if(other instanceof ListType)
			return other;
		return super.checkMultiply(context, other, tryReverse);
	}

	@Override
	public IType checkDivide(Context context, IType other) throws SyntaxError {
		if(other instanceof IntegerType)
			return DecimalType.instance();
		if(other instanceof DecimalType)
			return other;
		return super.checkDivide(context, other);
	}
	
	@Override
	public IType checkIntDivide(Context context, IType other) throws SyntaxError {
		if(other instanceof IntegerType)
			return this;
		return super.checkIntDivide(context, other);
	}

	@Override
	public IType checkModulo(Context context, IType other) throws SyntaxError {
		if(other instanceof IntegerType)
			return this;
		return super.checkModulo(context, other);
	}

	@Override
	public IType checkMember(Context context, Identifier name) throws SyntaxError {
		if(name.equals("min"))
			return this;
		else if(name.equals("max"))
			return this;
		else
			return super.checkMember(context, name);
	}

	@Override
	public IValue getMember(Context context, Identifier name) throws PrestoError {
		if(name.equals("min"))
			return new Integer(java.lang.Integer.MIN_VALUE);
		else if(name.equals("max"))
			return new Integer(java.lang.Integer.MAX_VALUE);
		else
			return super.getMember(context, name);
	}
	
	@Override
	public IType checkCompare(Context context, IType other) throws SyntaxError {
		if(other instanceof IntegerType)
			return BooleanType.instance();
		if(other instanceof DecimalType)
			return BooleanType.instance();
		return super.checkCompare(context, other);
	}
	
	@Override
	public IType checkRange(Context context, IType other) throws SyntaxError {
		if(other instanceof IntegerType)
			return new RangeType(this);
		return super.checkCompare(context, other);
	}
	
	@Override
	public Range<?> newRange(Object left, Object right) throws SyntaxError {
		if(left instanceof Integer && right instanceof Integer)
			return new IntegerRange((Integer)left,(Integer)right);
		return super.newRange(left, right);
	}

	@Override
	public ListValue sort(Context context, ICollection<IValue> list) throws PrestoError {
		return this.doSort(context,list);
	}
	
	@Override
	public IValue convertJavaValueToPrestoValue(Object value) {
        if (value instanceof Number)
            return new Integer(((Number)value).longValue());
        else
            return (IValue)value; // TODO for now
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/type/IntegerType.js
var NativeType = require("./NativeType").NativeType;
var BooleanType = require("./BooleanType").BooleanType;
var DecimalType = require("./DecimalType").DecimalType;
var CharacterType = null;
var ListType = require("./ListType").ListType;
var RangeType = require("./RangeType").RangeType;
var TextType = null;
var AnyType = require("./AnyType").AnyType;
var Integer = require("../value/Integer").Integer;
var IntegerRange = require("../value/IntegerRange").IntegerRange;
var PeriodType = null;

exports.resolve = function() {
	CharacterType = require("./CharacterType").CharacterType;
	TextType = require("./TextType").TextType;
	PeriodType = require("./PeriodType").PeriodType;
}

function IntegerType()  {
	NativeType.call(this, "Integer");
	return this;
}

IntegerType.prototype = Object.create(NativeType.prototype);
IntegerType.prototype.constructor = IntegerType;

IntegerType.instance = new IntegerType();
	
/*
public Class<?> toJavaClass() {
	return Long.class;
}
*/

IntegerType.prototype.isAssignableTo = function(context, other) {
	return (other instanceof IntegerType) || (other instanceof DecimalType) || (other instanceof AnyType);
};

IntegerType.prototype.checkAdd = function(context, other, tryReverse) {
	if(other instanceof IntegerType) {
		return this;
	} else if(other instanceof DecimalType) {
		return other;
	} else {
		return NativeType.prototype.checkAdd.call(this, context, other, tryReverse);
	}
};

IntegerType.prototype.checkSubstract = function(context, other) {
	if(other instanceof IntegerType) {
		return this;
	} else if(other instanceof DecimalType) {
		return other;
	} else {
		return NativeType.prototype.checkSubstract.call(this, context, other);
	}
};

IntegerType.prototype.checkMultiply = function(context, other, tryReverse) {
	if(other instanceof IntegerType) {
		return this;
	} else if(other instanceof DecimalType) {
		return other;
	} else if(other instanceof CharacterType) {
		return TextType.instance;
	} else if(other instanceof TextType) {
		return other;
	} else if(other instanceof PeriodType) {
		return other;
	} else if(other instanceof ListType) {
		return other;
	} else {
		return NativeType.prototype.checkMultiply.call(this, context, other, tryReverse);
	}
};


IntegerType.prototype.checkDivide = function(context, other) {
	if(other instanceof IntegerType || other instanceof DecimalType) {
		return DecimalType.instance;
	} else {
		return NativeType.prototype.checkDivide.call(this, context, other);
	}
};

IntegerType.prototype.checkIntDivide = function(context, other) {
	if(other instanceof IntegerType) {
		return this;
	} else {
		return NativeType.prototype.checkIntDivide.call(this, context, other);
	}
};

IntegerType.prototype.checkModulo = function(context, other) {
	if(other instanceof IntegerType) {
		return this;
	} else {
		return NativeType.prototype.checkModulo.call(this, context, other);
	}
};

IntegerType.prototype.checkMinus = function(context) {
	return this;
};

/*
@Override
public IType checkMember(Context context, String name) throws SyntaxError {
	if(name.equals("min"))
		return this;
	else if(name.equals("max"))
		return this;
	else
		return super.checkMember(context, name);
}

@Override
public IValue getMember(Context context, String name) throws PrestoError {
	if(name.equals("min"))
		return new Integer(java.lang.Integer.MIN_VALUE);
	else if(name.equals("max"))
		return new Integer(java.lang.Integer.MAX_VALUE);
	else
		return super.getMember(context, name);
}

*/

IntegerType.prototype.checkCompare = function(context, other) {
	if(other instanceof IntegerType || other instanceof DecimalType) {
		return BooleanType.instance;
	} else {
		return NativeType.prototype.checkCompare.call(this, context, other);
	}
};


IntegerType.prototype.checkRange = function(context, other) {
	if(other instanceof IntegerType) {
		return new RangeType(this);
	} else {
		return NativeType.prototype.checkRange.call(this, context, other);
	}
};


IntegerType.prototype.newRange = function(left, right) {
	if(left instanceof Integer && right instanceof Integer) {
		return new IntegerRange(left, right);
	} else {
		return NativeType.prototype.checkCompare.newRange(this, left, right);
	}
};

IntegerType.prototype.convertJavaScriptValueToPrestoValue = function(context, value, returnType) {
	if (typeof(value)=='number') {
		return new Integer(value);
	} else {
		return value; // TODO for now
	}
};


exports.IntegerType = IntegerType;
